### ISSUE TITLE
sparse option

### ISSUE THREAD

**Issue by yagui** [spyking-circus/spyking-circus#122]
In the converting step, if I use
sparse_export = True
all the cluster seems to be in the first 4 channels (I'm working with tetrodes and the radius is set to group channels by 4). 
![sparse](https://user-images.githubusercontent.com/4296015/59030608-6de3d900-882f-11e9-812f-592a8d20cf4d.png)
I have no problems adding the option 
export_all = True
to get all the unfitted spikes, but the result it is the same as the picture 

Using
sparse_export = False 
seems to work ok,
![no_sparse](https://user-images.githubusercontent.com/4296015/59030607-6de3d900-882f-11e9-84ec-2575d5be11de.png)

although if I also add
export_all = True

Then I get this error.

Traceback (most recent call last):
  File "/home/ariel/virtualenvs/spyking-test/.spyking/bin/circus-gui-python", line 11, in <module>
    load_entry_point('spyking-circus==0.8.0', 'console_scripts', 'circus-gui-python')()
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/circus/scripts/python_gui.py", line 194, in main
    gui = controller.create_gui()
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/apps/template/gui.py", line 797, in create_gui
    self.supervisor.attach(gui)
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 784, in attach
    gui=gui, sort=gui.state.get('ClusterView', {}).get('current_sort', None))
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 672, in _create_views
    data = [self._get_cluster_info(cluster_id) for cluster_id in self.clustering.cluster_ids]
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 672, in <listcomp>
    data = [self._get_cluster_info(cluster_id) for cluster_id in self.clustering.cluster_ids]
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 663, in _get_cluster_info
    out[key] = func(cluster_id)
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/utils/context.py", line 120, in memcached
    out = f(*args)
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/phy/apps/template/gui.py", line 228, in get_cluster_amplitude
    template_amplitudes = (data.max(axis=1) - data.min(axis=1)).max(axis=1)
  File "/home/ariel/virtualenvs/spyking-test/.spyking/lib/python3.6/site-packages/numpy/core/_methods.py", line 28, in _amax
    return umr_maximum(a, axis, None, out, keepdims, initial)
ValueError: zero-size array to reduction operation maximum which has no identity

The cluster list in phy shows only 

I'm using latest version at the moment
spyking-circus 07e95c9c86a7fc43fa0e6840ba7e921f5be298be
phy https://github.com/cortex-lab/phy/commit/e1a04fa194f9bb703ee9e3db6a5b08ea3dc853a6
phylib https://github.com/cortex-lab/phylib/commit/61bf68ba8176f91d3e09adacc7feff818cd951f9

**Comment by yger** [comment]
Can you send me the prb file? In fact, the sparse_export mode is mostly to speed up phy (when dealing with large probes) because then templates are internally used as sparse matrices. The other point is that by default, phy is colorcoding the 16 nearest channels next to the one where the template is peaking the most. This is annoying when you have shanks. So sparse_export should only colorcode channels where the template is defined.

**Comment by yagui** [comment]
I'm using this probe file for 8 tetrodes, trying to isolate each tetrode from the others.

-----------------------
total_nb_channels = 32
radius            = 5
channel_groups    = {1 : {}}

tetrodes = round(total_nb_channels/4)

channel_groups[1]["channels"] = list(range(total_nb_channels)) 
channel_groups[1]["graph"]    = []
channel_groups[1]["geometry"] = {}

for i in range(tetrodes):
  for j in range(4):
    channel_groups[1]["geometry"][i*4+j]= (0.0, i*10+j) 
----------------------------------

also it happens to me using the following probe file, with this sample data file
https://github.com/kwikteam/phy-data/blob/master/template/sim_binary.dat

-----------------------------------------
total_nb_channels = 34
radius            = 1.5

channel_groups = {
    1: {
        'channels': list(range(34)),
        'graph' : [],
        'geometry': {
              7: [ 1,  0],
              9: [ 0,  1],
             11: [ 0,  2],
             13: [ 1,  2],
             15: [ 0,  3],
             17: [ 1,  3],
             19: [ 0,  4],
             21: [ 1,  4],
             23: [ 0,  5],
             25: [ 1,  5],
             27: [ 0,  6],
             29: [ 1,  6],
             31: [ 0,  7],
              6: [ 1,  7],
              8: [ 0,  8],
             10: [ 1,  8],
             12: [ 1,  9],
             14: [ 0, 10],
             16: [ 1, 10],
             18: [ 0, 11],
             20: [ 1, 11],
             22: [ 0, 12],
             24: [ 1, 12],
             26: [ 0, 13],
             28: [ 1, 13],
             30: [ 0, 14],
              0: [ 1, 14],
              1: [ 0, 15],
              2: [ 1, 15],
              3: [ 0, 16],
              4: [ 1, 16],
              5: [ 0, 17],
             32: [ 0, 30],
             33: [ 1, 30]
        }
    }
}

**Comment by yagui** [comment]
The second issue, using sparse_export = False and export_all = True I think it is related to the 'amplitud' calculation that is done.

Using this example
https://github.com/kwikteam/phy-data/blob/master/template/sim_binary.dat
all the templates for the unfitted spikes has amplitud 0.

![amplitud](https://user-images.githubusercontent.com/4296015/59047993-f9229600-8852-11e9-8679-4c6d749d6b6c.png)
Although I don't know why when `sparse_export = True` they can be loaded, and in the other case, it fails.

**Comment by yger** [comment]
To reproduce the behavior, I'll download the file. But can you send me also the .params file?

**Comment by yagui** [comment]
I added .txt to be able to upload it
[sim_binary.params.txt](https://github.com/spyking-circus/spyking-circus/files/3266167/sim_binary.params.txt)

**Comment by yger** [comment]
I realized you are using phylib, so phy 2.0, am I rigth? I'm able to reproduce these bugs, but it may be phy 2.0 related, so I'll keep you updated

**Comment by yger** [comment]
The labelling bug for the sparse export mode (in the channel column) is now fixed in the latest version of phylib. I'll keep looking for the crash when setting export_all to True

**Comment by yagui** [comment]
I think this is related to the crash when exporting_all = True
The templates for the unfitted spikes or not shown in the template view (phy), and each of this spikes have amplitude 0 (which I think is what it makes the crash when using sparse_export = False).

In the following picture I selected one cluster of this unfitted spikes.
![a](https://user-images.githubusercontent.com/4296015/59364674-3aa8ba80-8d05-11e9-90a8-06e2fbb1b73b.png)

After I split the cluster in two, all the templates for unfitted spikes are shown in the template view, with the two new clusters.
![b](https://user-images.githubusercontent.com/4296015/59364779-6c218600-8d05-11e9-9bf0-510ad4939dcf.png)

If I undo the split, then I go back to the original template, but now the templates of unfitted spikes are shown
![c](https://user-images.githubusercontent.com/4296015/59364848-865b6400-8d05-11e9-8959-48dd861d2927.png)

In all this steps, the amplitude of the spikes remains zero.

**Comment by yger** [comment]
Keep in mind that this export_all option will give you all "unfitted" spikes, but we do not use much here, since we are trying to avoid as much as possible detouring clusters manually. I just pushed a patch in the master branch, since when converting, indeed amplitudes for unfitted spikes were set to 0. They should be 1 now if you reconvert the data

**Comment by yagui** [comment]
Yes amplitud is now 1, but with sparse_export = False I keep getting the same error, so I was wrong about that been the problem

```
Traceback (most recent call last):
  File "/home/ariel/virtualenvs/spyking/.spyking/bin/circus-gui-python", line 11, in <module>
    load_entry_point('spyking-circus==0.8.0', 'console_scripts', 'circus-gui-python')()
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/circus/scripts/python_gui.py", line 185, in main
    gui = controller.create_gui()
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/apps/template/gui.py", line 797, in create_gui
    self.supervisor.attach(gui)
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 776, in attach
    gui=gui, sort=gui.state.get('ClusterView', {}).get('current_sort', None))
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 664, in _create_views
    data = [self._get_cluster_info(cluster_id) for cluster_id in self.clustering.cluster_ids]
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 664, in <listcomp>
    data = [self._get_cluster_info(cluster_id) for cluster_id in self.clustering.cluster_ids]
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/cluster/supervisor.py", line 655, in _get_cluster_info
    out[key] = func(cluster_id)
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/utils/context.py", line 120, in memcached
    out = f(*args)
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/phy/apps/template/gui.py", line 228, in get_cluster_amplitude
    template_amplitudes = (data.max(axis=1) - data.min(axis=1)).max(axis=1)
  File "/home/ariel/virtualenvs/spyking/.spyking/lib/python3.6/site-packages/numpy/core/_methods.py", line 28, in _amax
    return umr_maximum(a, axis, None, out, keepdims, initial)
ValueError: zero-size array to reduction operation maximum which has no identity
```

**Comment by yger** [comment]
Give me few days to look into that. I think this is related to the fact
that phy is not working for templates defined on only a single channel. And
when you set export_all to True, the unfitted templates may be like that.
I'll look as soon as possible

Le mer. 12 juin 2019 à 19:15, Ariel Burman <notifications@github.com> a
écrit :

**Comment by yger** [comment]
I found the problem, and this has now been fixed, I think, for phy 2.0. The export_all method should work fine, irregardless of the sparse_export or not. This is now in the master branch, and the release 0.8.0 should arrive soon

**Comment by yagui** [comment]
I tested it and with my data it works ok.
The unfitted templates are shown in a strange way (with the latest phy(dev) and phylib) but it is not a big deal. Also I think is related to the amplitude column in the cluster view, where this unfitted templates have large amplitude numbers. 

![1](https://user-images.githubusercontent.com/4296015/59525295-f99adc80-8ea3-11e9-95bd-eeaed855aa23.png)

In the case of the sintetized dataset from the phy-data repository, I'm getting this error.
```
Traceback (most recent call last):
  File "/home/ariel/virtualenvs/spyking-dev/.spyking/bin/spyking-circus-subtask", line 11, in <module>
    load_entry_point('spyking-circus==0.8.0', 'console_scripts', 'spyking-circus-subtask')()
  File "/home/ariel/virtualenvs/spyking-dev/.spyking/lib/python3.6/site-packages/circus/scripts/subtask.py", line 41, in main
    circus.launch(task, filename, nb_cpu, nb_gpu, use_gpu, extension=extension)
  File "/home/ariel/virtualenvs/spyking-dev/.spyking/lib/python3.6/site-packages/circus/__init__.py", line 19, in launch
    module.main(params, nb_cpu, nb_gpu, use_gpu, extension)
  File "/home/ariel/virtualenvs/spyking-dev/.spyking/lib/python3.6/site-packages/circus/converting.py", line 279, in main
    N_tm = write_templates(output_path, params, extension)
  File "/home/ariel/virtualenvs/spyking-dev/.spyking/lib/python3.6/site-packages/circus/converting.py", line 130, in write_templates
    waveform = io.get_stas(params, times_i=spikes, mean_mode=True, neighs=[elec], src=elec, labels_i=np.ones(len(spikes)))
  File "/home/ariel/virtualenvs/spyking-dev/.spyking/lib/python3.6/site-packages/circus/shared/files.py", line 112, in get_stas
    local_chunk = numpy.dot(local_chunk, spatial_whitening)
ValueError: shapes (149,34) and (32,32) not aligned: 34 (dim 1) != 32 (dim 0)
```
